### ISSUE TITLE
Rust Language Server support

### ISSUE THREAD

**Issue by cetra3** [onivim/oni#1926]
Duplicate of https://github.com/onivim/oni/issues/1712 but I have tested this against the wiki config:

Tested with the following config.js:

```js
module.exports = {
    "language.rust.languageServer.command": "rls"
}
```

Does not work:

![screen shot 2018-03-26 at 3 03 07 pm](https://user-images.githubusercontent.com/6415435/37887028-d3e2e5b0-3106-11e8-9c1b-c22984c0b3cb.png)

`rls` is available on the path.

Also tested with absolute path:

```js
module.exports = {
    "language.rust.languageServer.command": "/Users/cetra/.cargo/bin/rls"
}
```

We get a bit further, but it still does not work:

![screen shot 2018-03-26 at 3 08 34 pm](https://user-images.githubusercontent.com/6415435/37887135-9cc3eed4-3107-11e8-91e5-68fc92040672.png)

**Comment by Drakulix** [comment]
I am using the following settings with success:
```
// Language Support - Rust
"language.rust.languageServer.command": "rustup",
"language.rust.languageServer.arguments": ["run", "stable", "rls"],
"language.rust.languageServer.rootFiles": ["Cargo.toml"],
```
I mostly selected them because `nightly` rust rls was crashing at the time, but maybe this modifies the environment or something? Most language-server plugins for rust, I know of, (atom and vscode) do use rustup to manage their rust installations, so I thought this might be of help.

**Comment by cetra3** [comment]
I have tried that command but I get the following messages in the console along with as pinnning wheel:

```
No grammar found for language: rust
Measure request fail
```

**Comment by appelgriebsch** [comment]
@Drakulix thanks for your snippets. Looks like Oni is trying to instantiate the RLS when using them, but still get the same problem as originally mentioned by the creator of the issue:

```
[Language Client] Notification textDocument/didOpen - P:\GH\rust\sample-rust\src\main.rs: start
bundle.js:1 [Language Client] Notification textDocument/didOpen - P:\GH\rust\sample-rust\src\main.rs: end
bundle.js:1 [LANGUAGE CLIENT - STDERR]: ERROR 2018-04-03T13:01:55Z: rls::server: dispatch error, Error { code: InvalidParams, message: "missing field `version`", data: None }

bundle.js:1 Received response message without id: Error is: 
{
    "code": -32602,
    "message": "missing field `version`"
}
```

**Comment by Drakulix** [comment]
@appelgriebsch I had the same issue in the beginning. I am trying to track down whats causing this and why I am currently not experiencing it anymore.

**Comment by gerd2002** [comment]
I am getting the same error here, any fix?

**Comment by wezm** [comment]
@Drakulix if you run `rustup run stable rls --version` what do you get?

**Comment by Drakulix** [comment]
@wezm `rls-preview 0.125.1-stable (cebf188 2018-03-19)`

**Comment by russoue** [comment]
I am using the following:

```
// Language Support - Rust
"language.rust.languageServer.command": "rustup",
"language.rust.languageServer.arguments": ["run", "nightly", "rls"],
"language.rust.languageServer.rootFiles": ["Cargo.toml"],
```

But nothing happens to me.  `rls` is in the path and works well with Atom.  Is there anything else I need to do to active Rust support on Oni?  Can it be that Oni is not recognizing the file name extension?

**Comment by Akin909** [comment]
@FredrikAleksander any luck successfully getting the `rls` to run without issue or is the above still happening?

**Comment by FredrikAleksander** [comment]
If I run the nightly version of RLS, it seems to work ok (sometimes diagnostics stops working temporarily, but it will start working again after a few changes). Looking at the RLS code, it was quite strict earlier and it would crash if any input was not what it expected (duplicate or out of sequence version numbers). In the nightly version it will just log a error instead

**Comment by appelgriebsch** [comment]
Interestingly I tested the Rust language server in a plain neovim environment with the LanguageClient-Neovim and this works quite well with the settings above (rustup run stable rls)...

**Comment by Akin909** [comment]
@appelgriebsch thanks for checking this with another client. I'm quite keen to get this working as well, I think our lsp client has some issues as it reports some errors with our current typescript client as well as our intended replacement `theia-ide-language-server` some clients seem to work well but others face issues as in #2448 I think we need to surface logs a bit more clearly and debug where what part of the client maybe causing the errors.

In your case can you get a screen shot of your dev console when the client fails to see what the error is?

**Comment by DWagenk** [comment]
I had similar problems, but after updating rls it now seems to work.
```
$ rustup run stable rls -V
rls-preview 0.127.0-stable (d2ade31 2018-05-08)
```
I'm a little confused, because it didn't work on the first try after updating, but only after the 4th or 5th reload of oni. I can't recall whether the error message was gone right after updating.

**Comment by appelgriebsch** [comment]
I did an update and install of rls and its dependencies as mentioned here: https://github.com/rust-lang-nursery/rls, which gives me that
```

λ rustup run stable rls -V
rls-preview 0.127.0-stable (d2ade31 2018-05-08)
```

It still shows me the same error mentioned above:

![oni-rust-error](https://user-images.githubusercontent.com/6803419/43185840-57218810-8fed-11e8-86fd-96a4f6239aa7.png)

**Comment by alasher** [comment]
I'm also running into this issue, running oni release 0.3.6. Seeing similar issues as reported in this thread, and have rls working fine on Atom.

```
$ rustup run stable rls -V
rls-preview 0.127.0-stable (d2ade31 2018-05-08)
```

**Comment by adamski** [comment]
I'm just trying out Oni, and having some similar issues. Any ideas?
<img width="408" alt="Screenshot 2019-06-03 at 15 47 07" src="https://user-images.githubusercontent.com/14980/58810966-dc9a1a00-8616-11e9-899b-aba147140368.png">
